did_finder_rucio/src/rucio_did_finder/replica_distance.py

```diff
@@ -0,0 +1,159 @@
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+import logging
+import os
+
+from typing import List, Mapping, Optional, Tuple
+from socket import gethostbyname
+import math
+from functools import lru_cache
+import tempfile
+from urllib.parse import urlparse
+import geoip2.database
+import geoip2.errors
+
+logger = logging.getLogger('ReplicaDistanceService')
+
+
+def _haversine_distance(lat1: float, lon1: float, lat2: float, lon2: float):
+    ''' Assume inputs are in degrees; will convert to radians. Returns distance in radians '''
+    dellat = math.radians(lat2-lat1)
+    dellon = math.radians(lon2-lon1)
+    hav_theta = ((1-math.cos(dellat))/2 +
+                 math.cos(math.radians(lat1))*math.cos(math.radians(lat2))*(1-math.cos(dellon))/2)
+
+    return 2*math.asin(math.sqrt(hav_theta))
+
+
+@lru_cache
+def _get_distance(reader: Optional[geoip2.database.Reader],
+                  fqdn: str, my_lat: float, my_lon: float):
+    """
+    Determine angular distance between server at fqdn and (my_lat, my_lon).
+    If there is a failure of fdqn location lookup, will return pi
+    """
+    if reader is None:
+        return math.pi
+    try:
+        loc_data = reader.city(gethostbyname(fqdn)).location
+    except geoip2.errors.AddressNotFoundError as e:
+        logger.warning(f'Cannot geolocate {fqdn}, returning maximum distance.\nError: {e}')
+        return math.pi
+    site_lat, site_lon = loc_data.latitude, loc_data.longitude
+    if site_lat is None or site_lon is None:
+        return math.pi
+    return _haversine_distance(site_lat, site_lon, my_lat, my_lon)
+
+
+class ReplicaSorter(object):
+    _reader: Optional[geoip2.database.Reader] = None
+    # we keep the temporary directory around so it won't get randomly deleted by the GC
+    _tmpdir: Optional[tempfile.TemporaryDirectory] = None
+
+    def __init__(self, db_url_tuple: Optional[Tuple[str, bool]] = None):
+        """
+        Argument is an optional tuple of (URL, bool).
+        The URL is assumed to be a file to download; the bool indicates whether
+        it is ready to be used (True) or needs unpacking (False)
+        """
+        if db_url_tuple is None:
+            db_url_tuple = self.get_download_url_from_environment()
+        self._download_data(db_url_tuple)
+
+    def sort_replicas(self, replicas: List[str], location: Mapping[str, float]) -> List[str]:
+        """
+        Main method of this class.
+        replicas: list of strings which are the URLs for the replicas for a file
+        location: dict of the form {'latitude': xxx, 'longitude': yyy} where xxx and yyy are floats
+        giving the latitude and longitude in signed degrees
+        """
+        if not self._reader:
+            return replicas
+        if len(replicas) == 1:
+            return replicas
+        fqdns = [(urlparse(replica).hostname, replica) for replica in replicas]
+        distances = [(_get_distance(self._reader, fqdn, location['latitude'],
```

Comment:
This statement could be a lot simpler if we had a class to hold each FQDN, and `distance` could be a method on it. Then we could just use a sorted list

Reply:
disagree that it would be technically simpler. The "class" would just be the moral equivalent of a `namedtuple` and we would have to instruct the sorting of what key to sort on. I suppose it could make it really explicit what is happening here, though what we have here is a pretty common Python idiom.